table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/HMSShim.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.table.under.hive.util;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.thrift.TApplicationException;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+/**
+ * Implements a shim layer for hive metastore client.
+ **/
+public class HMSShim implements HiveCompatibility {
+  private final TServiceClient mClient;
+
+  /**
+   * Constructor for HMSShim.
+   *
+   * @param client another client as delegate
+   */
+  public HMSShim(IMetaStoreClient client) {
+    // Because RetryingMetaStoreClient is itself a proxy, we need to get to the base class
+    while (Proxy.isProxyClass(client.getClass())) {
+      InvocationHandler handler = Proxy.getInvocationHandler(client);
+      if (handler.getClass().isAssignableFrom(RetryingMetaStoreClient.class)) {
+        client = getField(handler, "base");
+        continue;
+      }
+      // Other handlers can be added here
+      throw new RuntimeException("Unknown proxy handler for IMetaStoreClient");
+    }
+    mClient = getField(client, "client");
```

Comment:
Is this safe? We are grabbing a field by name, of the hdfs client?

Reply:
If this constructor fails, it will be caught at the higher level and we would behave as if there is no compatibility layer. 
This reflection allows us to access the  thrift client and issue hive 2.2 compatible requests.

Reply:
FYI: other places where reflection is used, there is the risk of failed reflection as well. In those cases, we catch throwable and fall back to the default behavior of the base client.